File: object-oriented-programming/virtual-world-java/src/Game/Model/Plants/Guarana.java
package Game.Model.Plants;

import Game.Model.Organism;
import Game.Model.Species;
import Game.Model.World;

import java.awt.*;
import java.util.Random;

public class Guarana extends Plant {
    Random random = new Random();

    public Guarana(int x, int y, World world){
        super(Species.GUARANA, x, y, 0, world, 1);
    }

    public Guarana(World world){
        super(Species.GUARANA, 0, 0, 0, world, 1);
        int width = world.getWidth();
        int height = world.getHeight();

        int newX, newY;

        do {
            newX = random.nextInt(width);
            newY = random.nextInt(height);
        } while(!world.isFree(newY, newX));

        world.setOrganism(null, y, x);
        x = newX;
        y = newY;
        world.setOrganism(this, y, x);
    }

    public void collision(Organism other){
        world.report(other.getSpecies().displayName() + " ate Guarana and got its strength incremented by 3!");
        markDead();
        world.setOrganism(null, y, x);
        other.moveOrganism(y, x);
        other.incrementStrength(3);
    }

    public final Color draw(){
        return species.displayColor();
    }

    public Guarana sow(int y, int x){
        return new Guarana(x, y, world);
    }
}

File: object-oriented-programming/virtual-world-cpp/src/plants/Guarana.cpp
#include "Guarana.h"
#include "World.h"
#include "Species.h"
#include <iostream>

Guarana::Guarana(int x, int y, World* world)
    : Plant(Species::GUARANA, x, y, 0, world, 1) {}

Guarana::Guarana(World* world)
    : Plant(Species::GUARANA, 0, 0, 0, world, 1) 
    {
        int rows = world->getRows();
        int columns = world->getColumns();

        int newX, newY;

        do{
            newX = rand() % rows;
            newY = rand() % columns;
        } while(!world->isFree(newX, newY));

        world->setOrganism(nullptr, x, y);
        x = newX;
        y = newY;
        world->setOrganism(this, x, y);
    }

Guarana::~Guarana() {}

void Guarana::collision(Organism* other){
    world->report((speciesToString(other->getSpecies())) + " ate " + (speciesToString(getSpecies())));
    markDead();
    world->setOrganism(nullptr, x, y);
    other->moveOrganism(x, y);
    other->incrementStrength(3);
    world->report(speciesToString(other->getSpecies()) + " got its strength incremented by 3!");
}

char Guarana::draw() const {
    return 'U';
}

Guarana* Guarana::sow(int x, int y) const {
    return new Guarana(x, y, world);
}

Guarana* Guarana::loadFrom(World* world, std::istream &in) {
    int s_x, s_y, s_strength, s_initiative, s_age, s_breedCooldown, s_canBreed;
    in >> s_x >> s_y >> s_strength >> s_initiative >> s_age >> s_breedCooldown >> s_canBreed;
    Guarana* guarana = new Guarana(s_x, s_y, world);
    guarana->strength = s_strength;
    guarana->initiative = s_initiative;
    guarana->age = s_age;
    guarana->breedCooldown = s_breedCooldown;
    guarana->canBreed = bool(s_canBreed);
    return guarana;
}